invoices/invoiceregistry.go

```diff
@@ -1092,7 +1096,18 @@ func (i *InvoiceRegistry) notifyExitHopHtlcLocked(
 			updateDesc.State != nil
 
 		// Assign resolution to outer scope variable.
-		resolution = res
+		if cancelSet {
+			// If a cancel signal was set for the htlc set, we set
+			// the resolution as a failure with a total mismatch in
+			// the amounts. Something was wrong with this htlc, so
+			// we probably can't settle the invoice at all.
+			resolution = NewFailResolution(
+				ctx.circuitKey, ctx.currentHeight,
+				ResultHtlcSetTotalMismatch,
+			)
+		} else {
```

Comment:
Should we use `ResultCanceled` here instead of `ResultHtlcSetTotalMismatch`?

I think we're unnecessarily assuming at this point that the HTLC set was cancelled because of a "total mismatch".

Or perhaps, to be more accurate we should add a new `FailResolutionResult` state called `ResultCanceledHtlcSet` or similar?

Reply:
The reason I picked this existing failure resolution is because it is the one closest to what's actually happening.

If we have 10 htlcs, and 1 of them pays the wrong way (e.g wrong currency), we do have some form of total amount mismatch

I'm also not against adding a new failure reason

Reply:
I'm not sure we can just add a new failure reason. Don't they map to a failure code defined in the BOLTS? Not sure to be honest.

The only reason against using `ResultHtlcSetTotalMismatch` I can see is that the error is related to AMP. But we're not necessarily always doing AMP, could also just be MPP.

I'm also not sure if `ResultCanceled` is better, since that implies the user actually cancelling the invoice (rejecting an HTLC will leave the invoice open and it can still be settled with the correct HTLCs).

So my vote would be for `ResultAmountTooLow`.

But I think the approach is the correct one, we just need to decide on an error code :+1: